### ISSUE TITLE
min = 0 should have some tolerance

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#10477]
Attributes such as `x(min = 0)` mean that `x >= 0` at all times. However, it is well-known that enforcing equality constraints on Real numbers is not a sound numerical practice, because of inevitable rounding errors. For example, consider the following MWE:
```modelica
model TestAssertTolerance2
  Real x(start = 1e4, nominal = 1e4, min = 0);
  function f
    input Real x;
    input Real w;
    output Real y;
  algorithm
    y := abs(x)^w*sign(x);
  end f;
equation
  der(x) = -50*f(x,0.8);
end TestAssertTolerance2;
```
The differential equation in this system represents a positive dynamical system: if the initial condition is positive, the value of `x` always remains positive, though it asymptotically approaches zero. Hence, one can state `x(min = 0)`, which is a structural property that could be used, e.g., to simplify `inStream()` operators.

Unfortunately, due to numerical errors, one gets:
```
The following assertion has been violated at time 0.636000 ((x >= 0.0)) --> 
"Variable violating min constraint: 0.0 <= x, has value: -9.60893e-07"
```
This warning is quite annoying. If one really wanted `x` to never, ever become negative, one should write `x(min = eps)`, with, e.g., `eps = 1e-6` or `eps = 1e-9`, so that there is some margin for numerical errors. If I write `min = 0`, it means I also accept also a zero value. But then, some slack for inevitable numerical errors should be added.

I would therefore propose to translate x(min = 0) into a check that `x >= -1e-9*x.nominal` or something like that. Similarly for max attributes.

I think the same should also apply for asserts such as assert(x >= 0, "x should not be negative"), but maybe this is slightly more questionable.

@bernhardbachmann, @phannebohm, @AnHeuermann what do you think?

**Comment by bilderbuchi** [comment]
>  If one really wanted x to never, ever become negative, one should write x(min = eps), with, e.g., eps = 1e-6 or eps = 1e-9, so that there is some margin for numerical errors. If I write min = 0, it means I also accept also a zero value. But then, some slack for inevitable numerical errors should be added.

I thought that if one wanted x to never, ever become negative, one chooses a modelling approach to enforce that, like it's taught e.g. in the [bouncing ball example](https://mbe.modelica.university/behavior/discrete/bouncing/)?

I agree that the warning can be annoying, and I also have selectively added epsilons in (few) places where numerical noise would trigger that warning _and I know that the limit excursion is otherwise harmless_. 

However, if I'm using `x` with the strong assumption is that it remains nonnegative (e.g. because an algorithm reacts badly to negative values), I rely on `min=0` to assert that precondition. If my algorithm sees a tiny negative `x`, but the assert warning does not fire because some blanket limit extensions have silently been added in the background, that would just muddy the waters and make debugging Modelica more complicated than it already is.

I think the decision if a "soft"/extended limit around `min`/`max` is actually OK or not will have to remain with the modeller, and the modeller can then decide to manually add small extensions around the limits. It puts a _bit_ more burden on the modeller to come up with sensible epsilon values, but I think that is not a bad thing, as it also keeps the fact that the limits are wider present and visible.
Silently making `min=0` actually mean `min=0-some_magic_value` would, in my opinion, be a change for the worse.

**Comment by AnHeuermann** [comment]
I think we shouldn't change the behavior of the check for `min` and `max`. If the model states the variable `x` should be in `min <= x <= max` the runtime should check just that.
Numeric noise can be an issue, but that is just the way it is. In your example it indicates that the ODE solver is a bit imprecise and the user needs to have a closer look at the result and decide if it is good enough or not. Or change the model / the settings to get a better result that holds the minimum constrain.

**Comment by phannebohm** [comment]
> If one really wanted `x` to never, ever become negative, one should write `x(min = eps)`, with, e.g., `eps = 1e-6` or `eps = 1e-9`, so that there is some margin for numerical errors. If I write `min = 0`, it means I also accept also a zero value.

I have to agree with @bilderbuchi and @AnHeuermann. This might as well be read the other way: If one really wanted to permit tiny negative values, one should write `x(min = -eps)`.

> This warning is quite annoying.

Perhaps so. But rather than closing ones eyes on it, the problem should be dealt with if it is one. We could discuss if the runtime should deal with this on another level, e.g. integrator step size, but that's another question.

> Silently making `min=0` actually mean `min=0-some_magic_value` would, in my opinion, be a change for the worse.

Furthermore I'm not convinced the runtime is able to always decide on a good `some_magic_value`.

**Comment by casella** [comment]
> I thought that if one wanted x to never, ever become negative, one chooses a modelling approach to enforce that, like it's taught e.g. in the [bouncing ball example](https://mbe.modelica.university/behavior/discrete/bouncing/)?

The `when h < 0` statement generates events with a zero-crossing function `zc = h` or `zc = sign(h)`. The event time is usually found by bisection, with a certain finite precision. Then the differential equation is integrated until that time, with some numerical approximation. There is absolutely no guarantee that h cannot become slightly negative at the bouncing events. 

In fact, this is exactly what happens with that model if you run it in OpenModelica:
![immagine](https://user-images.githubusercontent.com/4233148/229233937-e70daedd-83f4-4711-86d6-59203262aae8.png)


> I agree with you that the warning can be annoying, and I also have selectively added epsilons in (few) places where numerical noise would trigger that warning _and I know that the limit excursion is otherwise harmless_.

> However, if I'm using `x` with the strong assumption is that it remains nonnegative (e.g. because an algorithm reacts badly to negative values), I rely on `min=0` to assert that precondition. If my algorithm sees a tiny negative `x`, but the assert warning does not fire because some blanket limit extensions have silently been added in the background, that would just muddy the waters and make debugging Modelica more complicated than it already is.

That's a good point. 

However, my counterargument is that I have an application where flow rates are computed. The flow rates are meant to be always positive (entering), then you can write m_flow(min = 0) for the flow variable in the stream connector. According to stream connector semantics, this means that the corresponding stream variable is never considered for inStream() computation.
If a small negative flow rate is computed because of numerical errors, that is harmless, because it won't have any appreciable impact on the simulation results. However, I get blaring red warnings. On the other hand, if I set `m_flow(min = -1e-6)`, then according to stream connetor semantics, the steam variable of this component would be considered for inStream() computations. So, this doesn't really work.

As I wrote, if you really, really want a quantity to never, ever become negative, you have to enforce a finite positive limit, because numerical errors are inevitable. For example, if a negative height broke the bouncing ball model for some reason (e.g. a square root), then you should write `when h < 1e-6 then...`

> I think the decision if a "soft"/extended limit around `min`/`max` is actually OK or not will have to remain with the modeller, and the modeller can then decide to manually add small extensions around the limits.

That doesn't always work, see above.

> It puts a _bit_ more burden on the modeller to come up with sensible epsilon values, but I think that is not a bad thing, as it also keeps the fact that the limits are wider present and visible. Silently making `min=0` actually mean `min=0-some_magic_value` would, in my opinion, be a change for the worse.

But then, what if you actually have an equation stating x = 0 in your model, and you get x = -3.452e-12? That's what happened in my original use case, that unfortunately I could not replicate in an MWE here. Does that mean that the solution is wrong? No, it isn't, it's just that you have to accept numerical errors as a fact of life in equation-based languages.

**Comment by casella** [comment]
> I think we shouldn't change the behavior of the check for `min` and `max`. If the model states the variable `x` should be in `min <= x <= max` the runtime should check just that. Numeric noise can be an issue, but that is just the way it is. In your example it indicates that the ODE solver is a bit imprecise and the user needs to have a closer look at the result and decide if it is good enough or not. Or change the model / the settings to get a better result that holds the minimum constraint.

Unfortunately this turns out not to be possible, you always get some random small negative numbers, because errors are inevitable in numerical solvers. And then you always get those stupid error messages.

**Comment by casella** [comment]
> I have to agree with @bilderbuchi and @AnHeuermann. This might as well be read the other way: If one really wanted to permit tiny negative values, one should write `x(min = -eps)`.

But that would change the stream connector semantics in a completely unwanted and unnecessary way. See above.

> Perhaps so. But rather than closing ones eyes on it, the problem should be dealt with if it is one. 

The problem is, in many cases you just can't. The equations are solved numerically, and when the solution is expected to be zero, it can as well be -1.23e-12. You can't do much about that.

> Furthermore I'm not convinced the runtime is able to always decide on a good `some_magic_value`.

The runtime does it all the time, whether you like it or not. Every time a variable happens to be an iteration value of a nonlinear system, our current default seek a precision of 1e-12*x.nominal and then just stop the Newton iterations. Is that a magic value? Yes, but we have to live with it.

**Comment by phannebohm** [comment]
I guess if the limit is really mandated by some expression like `sqrt(x)` then the simulation fails anyway, so maybe we could allow for some deviation.

How about this: when a variable violates the limit a little e.g. ` -1e-12*x.nominal < x-x.min < 0` we reprimand it, meaning we remember the violation and make `x` promise it won't do it again. If the next step doesn't get better we issue the warning, if it does get better we set the new error as reference for the step after that. When `x` is back on track we forget the violation.

This way we allow for numerical errors but we still force `x` to eventually obey its limits.

**Comment by casella** [comment]
> How about this: when a variable violates the limit a little e.g. ` -1e-12*x.nominal < x-x.min < 0` we reprimand it, meaning we remember the violation and make `x` promise it won't do it again.

How can you enforce that?

> If the next step doesn't get better we issue the warning, if it does get better we set the new error as reference for the step after that. When `x` is back on track we forget the violation.

This seems to me a bit over-engineered, without any guarantees to work. Why should x get back on track, eventually? There's no theoretical reason why it should

At the end of the day, there is no such thing as absolute precision with IEEE floating points numbers. There will always be some numerical error. I guess we have to live with it.

> This way we allow for numerical errors but we still force `x` to eventually obey its limits.

Again, no way to convince solvers to actually do that. See what the numerical errors are in the MWE I provided. Sometimes it's positive, sometimes it's negative, that's pure chance.

**Comment by phannebohm** [comment]
> My point is that when you run `TestAssertTolerance2` you should get _zero_ warnings. There is no reason to make the end user nervous about something going wrong, when the only thing that is "going wrong" is a fact of life in numerical computations, i.e., numerical errors due to finite precision of IEEE 754 double precision numbers.

Agreed. But how do we distinguish that from an actual violation? Simulating `TestAssertTolerance2` with `x.min = -1e-9*x.nominal` still gives me:
```
The following assertion has been violated at time 0.684000
((x >= -1e-05)) --> \"Variable violating min constraint: -1e-05 <= x, has value: -1.6672e-05\"
```

Changing the min attribute to `-1e-8*x.nominal` gets rid of all asserts for this model. Do we just choose `1e-8` based on this? It feels a bit large for numerical errors. But then again I'm not a modeler so what do I know. :sweat_smile:

> Would your less-strict-but-not-as-loose-as-mine new strategy produce a warning with that model? How can you guarantee that it won't?

I guess I can't.

**Comment by casella** [comment]
That's a good point. 

In some cases, the expected numerical error is due to the solution of linear or non-linear equations, where the precision is normally pretty tight (e.g. 1e-12 relative precision), unless the condition number of the Jacobian is really very large, like 1e9.

On other cases (like with my MWE), this is due to the tolerance of the ODE solver, which is normally much less tight, e.g. 1e-4 or 1e-6. 

The problem is, the code handling the assert has no idea about that 😞

**Comment by casella** [comment]
We could, somewhat arbitrarily, set a limit at, say, 1e-6*nominal. We report a warning about the violation anyway, but in case it is less than this threshold, we can add something like: "This is probably due to numerical approximation errors." to the warning, so as to make the end user a bit less nervous. What do you think?

**Comment by phannebohm** [comment]
We haven't talked about actually enforcing the limits, yet. When I add to `TestAssertTolerance2` the equation
```modelica
when x < 0 then
  reinit(x, 0);
end when;
```
I get no asserts and the simulation runs smoothly. Well, with one state event at `time=0.679714798639`. Instead of introducing a bunch of when-equations, we could set states to their min or max values just after the integrator step when they overshoot. Maybe that is also desired for algebraic variables?

For models that actually violate their constraints this would probably lead to weird behavior. So maybe we could combine this with the epsilon based on nominal value and only "reinit" if x is within that small window.

Or is that too radical?

**Comment by phannebohm** [comment]
> In some cases, the expected numerical error is due to the solution of linear or non-linear equations, where the precision is normally pretty tight (e.g. 1e-12 relative precision), unless the condition number of the Jacobian is really very large, like 1e9.
> 
> On other cases (like with my MWE), this is due to the tolerance of the ODE solver, which is normally much less tight, e.g. 1e-4 or 1e-6.
> 
> The problem is, the code handling the assert has no idea about that disappointed

Maybe not. We know if x is a state or an iteration variable at codegen so we might be able to choose the scaling factor accordingly.

**Comment by phannebohm** [comment]
BTW, I was searching the MLS for min/max attributes and found [RealType](https://specification.modelica.org/master/class-predefined-types-and-declarations.html#S9.p8) and [assert](https://specification.modelica.org/master/equations.html#S3.SS7.p1). I've quoted the relevant parts here:

> 4.9.1 Real Type
>
> The following is the predefined Real type:
>
> ```modelica
> type Real // Note: Defined with Modelica syntax although predefined
>   [...]
>   parameter RealType min = -Inf, max = +Inf; // Inf denotes a large value
>   [...]
> equation
>   assert(min <= ⟨value⟩ and ⟨value⟩ <= max, "Variable value out of limit");
> end Real;
> ```


> 8.3.7 assert
>
> An equation or statement of one of the following forms is an assertion:
>
> ```modelica
> assert(condition, message); // Uses level=AssertionLevel.error
> [...]
> ```
> [...]
> [[...], the third formal parameter has the name level and the default value AssertionLevel.error.]
>
> [...] If the condition evaluates to false, different actions are taken depending on the level input:
>
> - level = AssertionLevel.error: The current evaluation is aborted. The simulation may continue with another evaluation. If the simulation is aborted, message indicates the cause of the error.
> [Ways to continue simulation with another evaluation include using a shorter step-size, or changing the values of iteration variables.]
> [...]

It seems to me, since the assert in `RealType` has `AssertionLevel.error`, that we should actually never continue the simulation with values that are outside their min/max attributes.

**Comment by casella** [comment]
Interesting discussion so far 😄 

> Instead of introducing a bunch of when-equations, we could set states to their min or max values just after the integrator step when they overshoot. Maybe that is also desired for algebraic variables?
> 
> For models that actually violate their constraints this would probably lead to weird behavior. 

Exactly. What worries me is that if I write the wrong state equations and then this policy constrains the state to remain non-negative, I will never discover that the state equations were wrong in the first place. Very dangerous.

> So maybe we could combine this with the epsilon based on nominal value and only "reinit" if x is within that small window.

This could be reasonable. The problem is, we also have algebraic variables depending on the states, and we could have asserts on them as well. This would require to multiply the tolerance on states by the sensitivity of those algebraic varables w.r.t. the states. This may soon become quite complicated. Or maybe not, as long as there are not so many algebraic variables with min/max limits.

> Maybe it does. We know if x is a state or an iteration variable at codegen so we might be able to choose the scaling factor accordingly.

But then we also need to handle algebraic variables depending on those states, see above.

> It seems to me, since the assert in RealType has AssertionLevel.error, that we should actually never continue the simulation with values that are outside their min/max attributes.

I wonder if those who wrote that part of the specification had an idea of the consequences of what they wrote. Maybe we should open a discussion on the MLS issue tracker about this.
